### PR TITLE
8329223: Parallel: Parallel GC resizes heap even if -Xms = -Xmx

### PR DESCRIPTION
**Pull request by zhengyu123**
Parallel GC resizes old gen even when setting -Xms == -Xmx. 

During GC argument setup, it always set `MinOldSize = GenAlignment` and does not make any adjustment when `-Xms == -Xmx`, while does the adjustment for young gen [here](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shared/genArguments.cpp#L250)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8329223](https://bugs.openjdk.org/browse/JDK-8329223): Parallel: Parallel GC resizes heap even if -Xms = -Xmx (**Bug** - P3)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18539/head:pull/18539` \
`$ git checkout pull/18539`

Update a local copy of the PR: \
`$ git checkout pull/18539` \
`$ git pull https://git.openjdk.org/jdk.git pull/18539/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18539`

View PR using the GUI difftool: \
`$ git pr show -t 18539`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18539.diff">https://git.openjdk.org/jdk/pull/18539.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18539#issuecomment-2026315868)

### CODE REVIEW

File: src/hotspot/share/gc/parallel/psOldGen.cpp
Comment (ecki) [inline src/hotspot/share/gc/parallel/psOldGen.cpp:297]: Is that a merge conflict and are min/max_gen_size() alligned?

Comment (zhengyu123) [reply src/hotspot/share/gc/parallel/psOldGen.cpp]: Oops, that's a leftover. Thank you for pointing out.

**Comment by albertnetymk** [comment]
Can `MinOldSize` be set as `MinOldSize = MinHeapSize - MinNewSize;` on L279, right after `MaxOldSize`?

**Comment by zhengyu123** [comment]
> Can `MinOldSize` be set as `MinOldSize = MinHeapSize - MinNewSize;` on L279, right after `MaxOldSize`?

Then the value is likely reset here: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shared/genArguments.cpp#L290

**Comment by albertnetymk** [comment]
I meant setting `MinOldSize` based the `new + old = heap` invariant on L279 and removing all other writes to it.

**Comment by zhengyu123** [comment]
> I meant setting `MinOldSize` based the `new + old = heap` invariant on L279 and removing all other writes to it.

Sorry for replying late, I was on PTO with spotty network connection.

Applied the changes you suggested, I am getting following errors:
`[0.013s][warning][gc,ergo] Inconsistency between generation sizes and heap size, resizing the generations to fit the heap.`
`Error occurred during initialization of VM`
`GC triggered before VM initialization completed. Try increasing NewSize, current value 192K.`

**Comment by albertnetymk** [comment]
I see; seems that there can be conflicting constraints from min, init and max triplets. Maybe sth like:

```
  MinOldSize = MIN3(MaxOldSize,
                    InitialHeapSize - initial_young_size,
                    MinHeapSize - MinNewSize);
```

**Comment by zhengyu123** [comment]
Make sense, updated.

**Comment by albertnetymk** [comment]
> updated.

The diff is the same as before, just fyi.

**Review by albertnetymk — APPROVED** [review]
I wonder what others think.

/reviewers 2

File: src/hotspot/share/gc/shared/genArguments.cpp
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/genArguments.cpp:312]: Do you have a test case that requires this? I'd expect `MIN3` above to be enough.

Comment (zhengyu123) [reply src/hotspot/share/gc/shared/genArguments.cpp:312]: Without this, gtest fails on MacOSX (https://github.com/zhengyu123/jdk/actions/runs/8743941793/job/23997262102)

File: src/hotspot/share/gc/shared/genArguments.cpp
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/genArguments.cpp:311]: Should this be `initial_old_size = MinOldSize;`?

Comment (albertnetymk) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: It's mostly to be consistent with `initial_old_size = MaxOldSize;` a few lines above; if the cmdline `OldSize` is inconsistent with other flags, change `OldSize` to fit others.

Technically, when there's inconsistency in cmdline flags, change any of them is equally correct, IMO. In the long run, maybe it's possible to remove the `OldSize` flag, since one can achieve old-size via `InitialHeapSize` and `NewSize`.

Comment (albertnetymk) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: Maybe my prev msg was misleading... Reprint the code-snippet:

```c
    if (initial_old_size > MaxOldSize) {
      log_warning(gc, ergo)(...);
      initial_old_size = MaxOldSize;
    } else if (initial_old_size < MinOldSize) {
      log_waring(gc, ergo)(...);
      initial_old_size = MinOldSize;
    }
```
so that the two assignment to `initial_old_size` matches up, accounting for over/under flow.

Comment (albertnetymk) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: I see. Could you explain a bit more on what causes the failure? (It's not obvious to me why some combination of initial_old_size/MinOldSize can trigger `pthread_mutex_destroy(&mutex_)failed`.)

Comment (zhengyu123) [reply src/hotspot/share/gc/shared/genArguments.cpp]: > I see. Could you explain a bit more on what causes the failure? (It's not obvious to me why some combination of initial_old_size/MinOldSize can trigger `pthread_mutex_destroy(&mutex_)failed`.)

I am not really familiar with gtest framework and not sure why. I could not reproduce the failure on Linux, so I submit the patch to see what is going to come out of github test.

Comment (zhengyu123) [reply src/hotspot/share/gc/shared/genArguments.cpp]: github action fails on macosx aarch64 (the same kind of machine I use) as well. I am not sure if it is test or test framework failure, as this part of code is platform dependent.

@albertnetymk what's your suggestion?

Comment (albertnetymk) [reply src/hotspot/share/gc/shared/genArguments.cpp]: > github action fails on macosx aarch64 (the same kind of machine I use)

Could you debug the failure on your macosx-aarch64 machine to identify how `initial_old_size = MinOldSize` results into the failure? (The error msg in GHA is rather limited.) One can probably print out the value for all related flags, `NewSize`, `OldSize`, `InitialHeapSize`, etc for the problematic case, and launch `java` on the cmdline (without `gtest`) with those flags.

> I could not reproduce the failure on Linux

My linux works fine as well, just fyi.

Comment (zhengyu123) [reply src/hotspot/share/gc/shared/genArguments.cpp]: Debugging with following command line:
`make CONF=macosx-aarch64-server-fastdebug run-test TEST=gtest TEST_VM_OPTS="-Xlog:gc+heap=trace"`

With `initial_old_size = MinOldSize` fails this test:

``` 
  // If OldSize is set on the command line, it should be used
  // for both min and initial old size if less than min heap.
  TestGenCollectorPolicy::SetOldSizeCmd setter(20 * M);

  TestGenCollectorPolicy::CheckOldMin checker_min(20 * M);
  TestGenCollectorPolicy::TestWrapper::test(&setter, &checker_min);
```

Debug output:
```
[0.001s][trace][gc,heap]   Maximum heap size 8589934592
[0.001s][trace][gc,heap]   Initial heap size 536870912
[0.001s][trace][gc,heap]   Minimum heap size 6815736
[0.008s][debug][gc,heap] Minimum heap 8388608  Initial heap 536870912  Maximum heap 8589934592
[0.036s][debug][gc,heap] Minimum heap 41943040  Initial heap 109051904  Maximum heap 8589934592
[0.036s][trace][gc,heap] 1: Minimum young 196608  Initial young 36306944  Maximum young 83886080
[0.036s][warning][gc,ergo] Inconsistency between initial old size and minimum old size
[0.036s][warning][gc,ergo] Inconsistency between generation sizes and heap size, resizing the generations to fit the heap.
[0.036s][trace  ][gc,heap] 2: Minimum young 196608  Initial young 67305472  Maximum young 83886080
[0.036s][trace  ][gc,heap] Minimum old 41746432  Initial old 41746432  Maximum old 8506048512
/Users/zhengyu.gu/ws/jdk/test/hotspot/gtest/gc/shared/test_collectorPolicy.cpp:261: Failure

```

With `MinOldSize = initial_old_size`

Debug output:

```
[0.001s][trace][gc,heap]   Maximum heap size 8589934592
[0.001s][trace][gc,heap]   Initial heap size 536870912
[0.001s][trace][gc,heap]   Minimum heap size 6815736
[0.007s][debug][gc,heap] Minimum heap 8388608  Initial heap 536870912  Maximum heap 8589934592
[0.034s][debug][gc,heap] Minimum heap 41943040  Initial heap 109051904  Maximum heap 8589934592
[0.034s][trace][gc,heap] 1: Minimum young 196608  Initial young 36306944  Maximum young 83886080
[0.034s][warning][gc,ergo] Inconsistency between initial old size and minimum old size
[0.034s][warning][gc,ergo] Inconsistency between generation sizes and heap size, resizing the generations to fit the heap.
[0.034s][trace  ][gc,heap] 2: Minimum young 196608  Initial young 83886080  Maximum young 83886080
[0.034s][trace  ][gc,heap] Minimum old 20971520  Initial old 25165824  Maximum old 8506048512

```

I will reverse back to `MinOldSize = initial_old_size`

Comment (albertnetymk) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: I can see these lines are diff:

```
[0.036s][trace  ][gc,heap] 2: Minimum young 196608  Initial young 67305472  Maximum young 83886080
[0.036s][trace  ][gc,heap] Minimum old 41746432  Initial old 41746432  Maximum old 8506048512
```

vs

```
[0.034s][trace  ][gc,heap] 2: Minimum young 196608  Initial young 83886080  Maximum young 83886080
[0.034s][trace  ][gc,heap] Minimum old 20971520  Initial old 25165824  Maximum old 8506048512
```

Does that mean using values from the problematic case, sth like `-XX:MinHeapSize=41943040 -XX:InitialHeapSize=109051904 -XX:MaxHeapSize=8589934592 -XX:NewSize=67305472 -XX:MaxNewSize=83886080 -XX:OldSize=20m`, one would get failure on launching java from cmdline on macosx-aarch64? (These values from the log are printed after they are adjusted as I read the code, so the actual cmdline flag values might not be diff. If you can reproduce the failure consistently on your box, further instrumenting might be required.)

It's still unclear to how the failure is triggered. Therefore, I can't say whether `MinOldSize = initial_old_size` fixes the issue, or just hides it.

Comment (zhengyu123) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: No, I believe it is the weirdness of the gtest. If I run from command line, both yield the same results:

`initial_old_size = MinOldSize`
```
MREM54BA45D6:hotspot zhengyu.gu$ $TEST_JDK/bin/java -XX:MinHeapSize=41943040 -XX:InitialHeapSize=109051904 -XX:MaxHeapSize=8589934592 -XX:NewSize=67305472 -XX:MaxNewSize=83886080 -XX:OldSize=20m -XX:+UseParallelGC -Xlog:gc+ergo,gc+heap=trace -version
[0.011s][debug][gc,heap] Minimum heap 41943040  Initial heap 109051904  Maximum heap 8589934592
[0.011s][trace][gc,heap] 1: Minimum young 41418752  Initial young 67108864  Maximum young 83886080
[0.011s][warning][gc,ergo] Inconsistency between generation sizes and heap size, resizing the generations to fit the heap.
[0.011s][trace  ][gc,heap] 2: Minimum young 41418752  Initial young 83886080  Maximum young 83886080
[0.011s][trace  ][gc,heap] Minimum old 524288  Initial old 25165824  Maximum old 8506048512
openjdk version "23-internal" 2024-09-17
OpenJDK Runtime Environment (fastdebug build 23-internal-adhoc.zhengyugu.jdk)
OpenJDK 64-Bit Server VM (fastdebug build 23-internal-adhoc.zhengyugu.jdk, mixed mode, sharing)
MREM54BA45D6:hotspot zhengyu.gu$ $TEST_JDK/bin/java  -XX:OldSize=20m -XX:+UseParallelGC -Xlog:gc+ergo,gc+heap=trace -version
[0.001s][trace][gc,heap]   Maximum heap size 8589934592
[0.001s][trace][gc,heap]   Initial heap size 536870912
[0.001s][trace][gc,heap]   Minimum heap size 22334664
[0.010s][debug][gc,heap] Minimum heap 25165824  Initial heap 536870912  Maximum heap 8589934592
[0.010s][trace][gc,heap] 1: Minimum young 1572864  Initial young 515899392  Maximum young 2863136768
[0.010s][trace][gc,heap] Minimum old 20971520  Initial old 20971520  Maximum old 5726797824
openjdk version "23-internal" 2024-09-17
OpenJDK Runtime Environment (fastdebug build 23-internal-adhoc.zhengyugu.jdk)
OpenJDK 64-Bit Server VM (fastdebug build 23-internal-adhoc.zhengyugu.jdk, mixed mode, sharing)
```

`MinOldSize = initial_old_size;`
```
MREM54BA45D6:hotspot zhengyu.gu$ $TEST_JDK/bin/java -XX:MinHeapSize=41943040 -XX:InitialHeapSize=109051904 -XX:MaxHeapSize=8589934592 -XX:NewSize=67305472 -XX:MaxNewSize=83886080 -XX:OldSize=20m -XX:+UseParallelGC -Xlog:gc+ergo,gc+heap=trace -version
[0.014s][debug][gc,heap] Minimum heap 41943040  Initial heap 109051904  Maximum heap 8589934592
[0.014s][trace][gc,heap] 1: Minimum young 41418752  Initial young 67108864  Maximum young 83886080
[0.014s][warning][gc,ergo] Inconsistency between generation sizes and heap size, resizing the generations to fit the heap.
[0.014s][trace  ][gc,heap] 2: Minimum young 41418752  Initial young 83886080  Maximum young 83886080
[0.014s][trace  ][gc,heap] Minimum old 524288  Initial old 25165824  Maximum old 8506048512
openjdk version "23-internal" 2024-09-17
OpenJDK Runtime Environment (fastdebug build 23-internal-adhoc.zhengyugu.jdk)
OpenJDK 64-Bit Server VM (fastdebug build 23-internal-adhoc.zhengyugu.jdk, mixed mode, sharing)
MREM54BA45D6:hotspot zhengyu.gu$ $TEST_JDK/bin/java  -XX:OldSize=20m -XX:+UseParallelGC -Xlog:gc+ergo,gc+heap=trace -version
[0.002s][trace][gc,heap]   Maximum heap size 8589934592
[0.002s][trace][gc,heap]   Initial heap size 536870912
[0.002s][trace][gc,heap]   Minimum heap size 22334664
[0.015s][debug][gc,heap] Minimum heap 25165824  Initial heap 536870912  Maximum heap 8589934592
[0.015s][trace][gc,heap] 1: Minimum young 1572864  Initial young 515899392  Maximum young 2863136768
[0.015s][trace][gc,heap] Minimum old 20971520  Initial old 20971520  Maximum old 5726797824
openjdk version "23-internal" 2024-09-17
OpenJDK Runtime Environment (fastdebug build 23-internal-adhoc.zhengyugu.jdk)
OpenJDK 64-Bit Server VM (fastdebug build 23-internal-adhoc.zhengyugu.jdk, mixed mode, sharing)
```

**Review by albertnetymk — APPROVED** [review]
It's the failing assert in `CheckOldMin` that is causing the `pthread_mutex_destroy` error msg. (I can recreate the error msg on my linux after adding a `ASSERT_LT(2, 1)` there.)

Then, I think the current fix should be safe, and keeps existing tests happy.

The longer term fix should probably be removing `OldSize`...

**Comment by zhengyu123** [comment]
Friendly ping! Could I get a second review.

File: src/hotspot/share/gc/shared/genArguments.cpp
Comment (lgxbslgx) [inline src/hotspot/share/gc/shared/genArguments.cpp:311]: The comment shown below seems strange. Can it be removed?
```C++
    // The generation minimums and the overall heap minimum should
    // be within one generation alignment.
```


The warning message shown below can be more consistent with this newly added one. 
```C++
      log_warning(gc, ergo)("Inconsistency between maximum heap size and maximum "
                            "generation sizes: using maximum heap = " SIZE_FORMAT
                            ", -XX:OldSize flag is being ignored",
                            MaxHeapSize);
```

**Review by lgxbslgx — CHANGES_REQUESTED** [review]
Looks good. Two nits/questions.

File: src/hotspot/share/gc/shared/genArguments.cpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: If you think they are not related to this bug, it is also OK to fix them in another ticket.

Comment (zhengyu123) [reply src/hotspot/share/gc/shared/genArguments.cpp:311]: @lgxbslgx Thanks for the review, filed [JDK-8331432](https://bugs.openjdk.org/browse/JDK-8331432)

**Review by lgxbslgx — APPROVED** [review]

**Comment by zhengyu123** [comment]
Thanks, @albertnetymk and @lgxbslgx

**Pull request closed without merging**
